runtime/compiler/ilgen/Walker.cpp

```diff
@@ -5061,13 +5061,54 @@ TR_J9ByteCodeIlGenerator::loadAuto(TR::DataType type, int32_t slot, bool isAdjun
    push(load);
    }
 
+static bool isFieldResovled(TR::Compilation *comp, TR_ResolvedJ9Method * owningMethod, int32_t cpIndex, bool isStore)
+   {
+   uint32_t offset = 0;
+   TR::DataType type = TR::NoType;
+   bool isVolatile = true, isFinal = false, isPrivate = false, isUnresolvedInCP;
+   bool bResolved = owningMethod->fieldAttributes(comp, cpIndex, &offset, &type, &isVolatile, &isFinal,
+                                    &isPrivate, isStore, &isUnresolvedInCP, true /* needsAOTValidation */);
+   return bResolved;
+   }
+
 void
 TR_J9ByteCodeIlGenerator::loadInstance(int32_t cpIndex)
    {
    if (_generateReadBarriersForFieldWatch && comp()->compileRelocatableCode())
       comp()->failCompilation<J9::AOTNoSupportForAOTFailure>("NO support for AOT in field watch");
-   TR::SymbolReference * symRef = symRefTab()->findOrCreateShadowSymbol(_methodSymbol, cpIndex, false);
-   loadInstance(symRef);
+
+   TR_ResolvedJ9Method * owningMethod = static_cast<TR_ResolvedJ9Method*>(_methodSymbol->getResolvedMethod());
+
+   if (TR::Compiler->om.areValueTypesEnabled() &&
+       owningMethod->isFieldQType(cpIndex))
+      {
+      if (isFieldResovled(comp(), owningMethod, cpIndex, false))
+         {
+         if (owningMethod->isFieldFlattened(comp(), cpIndex, _methodSymbol->isStatic()))
```

Comment:
This sub-tree should have some tracing for when traceILGen is on to outline the choice made and why so we know why we generated the tree we did.

Reply:
The tracing is added to where each flattened field is being loaded.